src/EnergyPlus/PlantLoopHeatPumpEIR.cc

```diff
@@ -1292,4 +1293,1368 @@ void EIRPlantLoopHeatPump::oneTimeInit(EnergyPlusData &state)
+            ShowContinueError(state, format(" EIR Modifier curve (function of Temperatures) output is negative ({:.3T}).", eirModifierFuncTemp));
+            ShowContinueError(state,
+                              format(" Negative value occurs using a water temperature of {:.2T}C and an outdoor air temperature of {:.2T}C.",
+                                     waterTempforCurve,
+                                     oaTempforCurve));
+            ShowContinueErrorTimeStamp(state, " Resetting curve output to zero and continuing simulation.");
+        }
+        ShowRecurringWarningErrorAtEnd(state,
+                                       format("{} \"{}\": EIR Modifier curve (function of Temperatures) output is negative warning continues...",
+                                              DataPlant::PlantEquipTypeNames[static_cast<int>(this->EIRHPType)],
+                                              this->name),
+                                       this->eirModFTErrorIndex,
+                                       eirModifierFuncTemp,
+                                       eirModifierFuncTemp);
+        eirModifierFuncTemp = 0.0;
+    }
+
+    Real64 miniPLR_mod = this->minPLR; // 0.25; // 2022-05-17: maybe should use input minPLR; however, this is to duplicate the ems verson
+    Real64 PLFf = max(miniPLR_mod, partLoadRatio);
+
+    Real64 eirModifierFuncPLR = Curve::CurveValue(state, this->powerRatioFuncPLRCurveIndex, PLFf);
+    // this->powerUsage = (this->loadSideHeatTransfer / this->referenceCOP) * eirModifierFuncPLR * eirModifierFuncTemp;
+    // this->powerEnergy = this->powerUsage * reportingInterval;
+
+    if (eirModifierFuncPLR < 0.0) {
+        if (this->eirModFPLRErrorIndex == 0) {
+            ShowSevereMessage(state, format("{} \"{}\":", DataPlant::PlantEquipTypeNames[static_cast<int>(this->EIRHPType)], this->name));
+            ShowContinueError(state, format(" EIR Modifier curve (function of PLR) output is negative ({:.3T}).", eirModifierFuncPLR));
+            ShowContinueError(state, format(" Negative value occurs using a Part Load Ratio of {:.2T}", PLFf));
+            ShowContinueErrorTimeStamp(state, " Resetting curve output to zero and continuing simulation.");
+        }
+        ShowRecurringWarningErrorAtEnd(state,
+                                       format("{} \"{}\": EIR Modifier curve (function of PLR) output is negative warning continues...",
+                                              DataPlant::PlantEquipTypeNames[static_cast<int>(this->EIRHPType)],
+                                              this->name),
+                                       this->eirModFPLRErrorIndex,
+                                       eirModifierFuncPLR,
+                                       eirModifierFuncPLR);
+        eirModifierFuncPLR = 0.0;
+    }
+
+    Real64 oaTemp2 = max(-8.8888, min(3.3333, oaTempforCurve));
```

Comment:
Anywhere we are doing min(max(a, b), c) or the reverse should move to using `std::clamp` as we go forward.  @amirroth that's for all applications, right?  Or is there something specific we should look for?

Reply:
Has the NAN stuff been worked out?

Reply:
There is a branch where @jmarrec has addressed many of them, but that's just a problem within the unit tests.  When I changed to std::clamp, it revealed that a few unit tests were missing a variable initialization and calculating NaN, which was getting masked by the min(max( chain.  It shouldn't be a problem at all in our actual running code, so changing to std::clamp should be just fine.

I really just wondered if there was any special cases where we wouldn't want to do clamp.  If it's just a universal improvement, then I'll just always ask for it.  :+1:

Reply:
IMHO, the only thing to consider is that it's technically undefined behavior if you call `std::clamp(v, lo, hi) `when `lo` is greater than `hi`.

```
Real64 const V5 = std::clamp(Var5, thisCurve->inputLimits[4].min, thisCurve->inputLimits[4].max);
```

Eg you gotta make sure beforehand than the input min/max are correctly ordered. But using `std::max(lo, std::min(hi, v))` has the same issue.

They may end up returning a different value depending on inputs but it'll be wrong to begin with.

here's a test bed for fun: https://godbolt.org/z/vG3Pe3djj


FYI, GCC's implementation does an assert ` __glibcxx_assert(!(__hi < __lo));`, so any problem will be caught when:
* it's constant evaluated (`constexpr double d = std::clamp(10., 18., 12.)`) , or,
* if `_GLIBCXX_ASSERTIONS` is 1, which is the case when `_GLIBCXX_DEBUG` is defined. 

**Our GCC Debug build does define it , so it'll catch these mistakes, which makes std::clamp much preferable to me** (in addition to readability)

https://github.com/NREL/EnergyPlus/blob/431463a7464daf007113f93c984a263a970b54bb/cmake/CompilerFlags.cmake#L97


`/usr/include/c++/11/bits/stl_algo.h`

```c++
#if __cplusplus > 201402L

#define __cpp_lib_clamp 201603

  /**
   *  @brief  Returns the value clamped between lo and hi.
   *  @ingroup sorting_algorithms
   *  @param  __val  A value of arbitrary type.
   *  @param  __lo   A lower limit of arbitrary type.
   *  @param  __hi   An upper limit of arbitrary type.
   *  @return max(__val, __lo) if __val < __hi or min(__val, __hi) otherwise.
   */
  template<typename _Tp>
    constexpr const _Tp&
    clamp(const _Tp& __val, const _Tp& __lo, const _Tp& __hi)
    {
      __glibcxx_assert(!(__hi < __lo));
      return (__val < __lo) ? __lo : (__hi < __val) ? __hi : __val;
    }
#endif // C++17
```

Reply:
> Our GCC Debug build does define it , so it'll catch these mistakes, which makes std::clamp much preferable

That's fine for cases that are clamping between hard values of 0.0 and 1.0, for example, but curve limits are user inputs, so we need to make sure those are checked (and fixed or fataled) during getinput.

Reply:
A few places in the code were modified to use `std::clamp(&val, &min, &max, opt_comp)`. In the original `std::clamp` documentation, I do not think I have fully understood why it is defined using reference type inputs and also return a reference type value (probably due to the class type template generalization?)---actually it looks to me that they would be much easier to understand when using value-based parameter transfers, at least for the primitive types (like `double`). Nevertheless, I think it probably would not affect using it in the few scenarios occurred in the current code.

Reply:
🤔 That's a very good point. I didn't notice that and you definitely don't want to create references unnecessarily! My suggestion is that we create a `constexpr` function `rclamp` (or `dclamp` or something) in `General` or `Utilities` or something and use that instead of `std::clamp`. Don't hold up this PR for that, but that could be a quick pass afterwards.